pythonForPilots - Lesson 1 - The Basics
==============

## Something to Keep in Mind up Front
I think one thing that's worth saying right up front is that **everything in Python is an object**.

This may not mean anything to you right now, but it's just kind of something to keep in the back of your head. But for now, just know it means that Python provides *alot* of flexibility.

And flexibility is good.

## Comments
I'm just going to start with how to insert comments real quick since, well, comments are quite important for readable code.

There is only one real way of creating comments:

In [1]:
# Inline comments are preceeded with a #
1 == 1  # Inline comments can be included after code as well

True

There is unfortunately no real way to do block comments. You might see something like the following, but this is considered a [docstring](https://www.python.org/dev/peps/pep-0008/#documentation-strings):

In [2]:
def foo():
    """
    This function does absolutely nothing
    Except prove a point
    :returns: None
    """
    pass

In reality, these triple double-quoted blocks are actually strings (which I'll get to later).

## Whitespace
Whitespace in Python is used to denote code blocks, such as for function definitions or loops. The previous example showing the docstring example should be a good example.

One thing to note is that it doesn't matter if you indent with tabs or spaces, but you need to be consistent

## Variables and Operators
### Naming and Assignment
Varable names are case sensitive and must start with either a letter or an underscore. After that, they can contain any combintation of alphanumeric characters as well as underscores.

Assignments are straight forward as well. Let's take a look:

In [3]:
aircraft1_make = 'Diamond'
FleetSize = 20
airline_codes = ['9E', 'OO', 'ZW', 'XJ']
aircraft_codes = {'DV20': 'Diamond DA20', 'C172': 'Cessna 172'}

As you can see, it's nothing too crazy. If you don't understand the types with the last two assignments, don't worry, I'll be covering those in a minute.

### Operators
Python has a numerous list of operators. For a list, I'm going to send you [here](https://www.programiz.com/python-programming/operators) instead of the actual Python documentation because, well, it's going to be more straight forward (if you're curious, actual documentation is [here](https://docs.python.org/3/reference/expressions.html)).

In [4]:
pax_count = 100
new_pax_count = pax_count + 5

aircraft1_make = aircraft1_make + ' Aircraft'

print(pax_count)
print(new_pax_count)
print(aircraft1_make)
print(True or False)
print(not True)
print('On ' + 'the ' + 'fly')

100
105
Diamond Aircraft
True
False
On the fly


The example above should be pretty straight forward with the `pax_count` addition. But you'll also notice that strings can be concatenated with `+` as well.

But you'll also notice the `or` and `not` operators. `or` is a logical comparison operator that checks for a value of `True` on either side, and `not` basically inverts the boolean following it.

### Augmented Assignments
Beyond the simple assignment operator (`=`), you can also do augmented assignments. Instead of explaining, I think a simple demo will get the point accross:

In [5]:
print('Current Fleet Size: ' + str(FleetSize))

FleetSize *= 5  # Apparently something good happened

print('New Fleet Size: ' + str(FleetSize))

Current Fleet Size: 20
New Fleet Size: 100


Take a look at the list of operators for available augmented assignment operators, but as you can see, they can shorten up situations where you want to use the same variable name after a simple operation.

### Object Mutability
_Credit to this [article](https://medium.com/@meghamohan/mutable-and-immutable-side-of-python-c2145cf72747) for helping me explain this subject._

One thing that is important to know is that some object types are mutable (can be changed) while others are immutable (can't be changed).

Some built-in types that are immutable include:

* `str` (Strings)
* `int` (Integers)
* `float` (Floating Point Numbers)

Some built-in types that are mutable include:

* `list` (Lists)
* `dict` (Dictionaries)
* `set` (Sets)

Additionally, custom object types are generally mutable.

Again, don't worry if you don't understand some of the types above - I'll get to those momentarily.

Before going further, I'd like to address one thing: Variables reference objects. When `FleetSize` changed from 20 to 100, the actual underlying object that `FleetSize` referenced (20) didn't change - but rather, it referenced a different object (100).

I think a demonstration should clear this up, while demonstrating mutability vs immutability:

In [6]:
# Looking at the object identities for immutable objects
print(id(FleetSize))
print(id(100))

print('\n')

FleetSize = 5
print(id(FleetSize))
print(id(5))

140736750264200
140736750264200


140736750261160
140736750261160


In the most basic sense, `id` function returns the identity of an object. As you can see, when `FleetSize` referenced the object represtning an integer with the value of 100. After `FleetSize` was assigned an integer with the value of 5, the object (as indicated by it's ID) was different.

Now let's take a look at what happens with a mutable object (a list):

In [7]:
print(airline_codes)
print(id(airline_codes))

airline_codes.append('WN')

print(airline_codes)
print(id(airline_codes))

['9E', 'OO', 'ZW', 'XJ']
1854866713856
['9E', 'OO', 'ZW', 'XJ', 'WN']
1854866713856


And for good measure...

In [8]:
airline_codes_part_deux = airline_codes
print(id(airline_codes_part_deux))

airline_codes_part_deux.append('YV')
print(airline_codes)

1854866713856
['9E', 'OO', 'ZW', 'XJ', 'WN', 'YV']


And one final opportunity to screw with your head:

In [9]:
print(id(['9E', 'OO', 'ZW', 'XJ', 'WN', 'YV']))
print(id(['9E', 'OO', 'ZW', 'XJ', 'WN', 'YV']))

1854866718976
1854866719168


As you can see, even though a new code was added to `airline_codes`, it maintained the same object ID. Additionally, running a mutation method against `airline_codes_part_deux` and then subsequently printing the value of the original `airline_codes` emphasizes the point about variables being a reference to an object.

Then the last example showed that when these mutable objects are created, even with identical values, they are in fact different objects.

## Basic Data Types
Now that we've gone over operators and variable assignments, let's take a look at some of the built-in types (also known as classes, which I may use interchangably). For a comprehensive list, take a look at the [docs](https://docs.python.org/3/library/stdtypes.html).

I'm not going to go horrendously indepth into all the things you can do with these types, but rather go over the common things you can do with them for a basic understanding (this is Lesson 1, after all).

### Strings
Strings are, as the documentation says, "immutable sequences of Unicode code points". If you're going through these lessons, you're using Python 3.7. Don't worry about the bit mentioning "Unicode code points". Python 2 strings were ASCII, but that's all I'll say on the matter here. For now, just know strings are essentially a sequence of characters.

#### Creation
Let's look at the two basic ways to create strings:

In [10]:
callsign = 'Flagship'
flight_number = str(3701)

You have obviously already seen the first way of creating a string by using single quotes. You can also use double (`"String"`) or triple quotes (`'''String'''` or `"""String"""`). Using single quotes allows you to use double quotes within the string without needing any escaping, and vice-versa. Triple quoting allows you to run the string over multiple lines. This method also includes all the whitespace as well, so that's something to keep in mind.

You might be wondering why you would use the second way. Well, let's say we wanted to append the flight number to the callsign:

In [11]:
try:
    print(callsign + ' ' + 3701)
except TypeError as err:
    print(err)

can only concatenate str (not "int") to str


Whupz. As you can see, Python doesn't allow you to concatenate a string with another object type. This is why getting the string representation of other object types is required first - most commonly by wrapping it in `str()`.

In [12]:
print(callsign + ' ' + flight_number)

Flagship 3701


Ahh, there we go.

#### String Slices
As previously mentioned, strings are essentially a sequence of characters. This means we can do stuff like:

In [13]:
print(aircraft1_make[0:7])  # Print the first seven characters
print(aircraft1_make[-8:])  # Print the last eight characters
print(aircraft1_make[::2])  # Print every other character
print(aircraft1_make[::-1]) # Print the string in reverse

Diamond
Aircraft
DaodArrf
tfarcriA dnomaiD


To start, sequences in Python are "Zero Indexed". This basically means that the index for the first element of a sequence, such as a string here, is 0. The following explanation of slices will also apply to other sequence types, such as lists.

These slices come in three forms:

* `[i]` - This grabs the value at index `i`. Again, remember the first element is indexed at 0.
* `[i:j]` - This grabs the values from `i` to `j`. Important to note that the element indexed at `j` is **not** included.
* `[i:j:k]` - This grabs the values ranged in `i` to `j`, but in the number of steps as in `k`.

Of course, as you can see above, indexes can also be negative (the last value of a sequence being `-1`), and values can be ommitted, as exemplified by the following method of printing an entire string that is Rube Goldberg approved:

In [14]:
print(aircraft1_make[::])

Diamond Aircraft


#### Formatting Strings
Let's say we wanted to be a little more... elegant in our printing of the above flight. Instead of having to resort to string concatenation `print(callsign + ' ' + flight_number)`, we can use string formatting. There are a few different ways to do this. One such way is like so:

In [15]:
print('{callsign} {number}'.format(callsign=callsign, number=flight_number))

Flagship 3701


The above uses named placeholders with the values specified by keywords. [This site](https://pyformat.info/) has nice list of the different ways you can do string formatting using the `format` method.

However, since python 3.6, _f-strings_ are another way to do string formatting:

In [16]:
print(f'{callsign} {flight_number}')

Flagship 3701


In this particular instance, you'll see that f-strings also use curley braces to identify things to get expanded, although in this case it's done in place without calling the `format` method.

#### Escaping
Before finishing up on strings, I should probably discuss escaping real quick. Let's say you wanted to print the following bit of text:
```
"Diamond 150PT, you're cleared for the option on Runway 13"

"Cleared for the option on 13, 150PT"
```

In this case, we both have single- and double-quotes. We also have a new line we want to insert. What do? Escape characters.

In [17]:
print('"Diamond 150PT, you\'re cleared for the option on Runway 13"\n')
print("\"Cleared for the option on 13, 150PT\"")

"Diamond 150PT, you're cleared for the option on Runway 13"

"Cleared for the option on 13, 150PT"


So as you can see, if you need to use a quote of the same type that it's being enclosed by, escape it with a backslash. `\n` will insert a newline.

### Numeric Types
Python has a couple different built-in numeric types: `int` (Integers), `float` (Floating point), and `complex` (Complex numbers). The constituents of the `complex` type are both floating point numbers.

One thing to note is that calculations on floating point numbers are not precise:

In [18]:
float_example = 1.1 + 2.2
print(float_example)

3.3000000000000003


If you're wondering why it's `3.3000000000000003` and not just `3.3`, you're going to have to ask someone smarter than me. If you do need exacting calculations with decimals, you'll need the `decimal` module:

In [19]:
from decimal import Decimal # We'll get to this later

decimal_example = Decimal('1.1') + Decimal('2.2')  # Notice these are entered as strings
print(decimal_example)

3.3


Again, don't worry about the module import stuff for now. Just know that if you need to do stuff with exacting precision, use the `Decimal` class. Also keep in mind that there are always tradeoffs - the calculations done will be significantly slower, although for our purposes here, not noticeable.

### Booleans and None
Booleans are just as you would imagine - representations of true and false. The literals are defined as `True` and `False`. Comparison operators evaluate to one of these two (not surprisingly).

In [20]:
print(True or False)
print(True and False)
print(1 < 2)
print(1 == 2)

True
False
True
False


As you can see, straight forward.

`None` can be viewed as the Python equivalent of a null. It's commonly used as a placeholder for variables, and subsequent checking if that variable was populated with something else like so:

In [21]:
successful_landing = None
crash = False
if successful_landing is not None:
    print('The landing was successful: {success}'.format(success=successful_landing))

if not crash:
    successful_landing = True
if successful_landing is not None:
    print('The landing was successful: {success}'.format(success=successful_landing))

The landing was successful: True


Contrived example? Absolutely. Did it get the point across? I hope so.

### Lists
Just like strings, lists contain a sequence of objects - except the objects can be of any type. You've seen them in the earlier examples.

#### Instantiation
Lists are instantiated like so:

In [22]:
airline_codes = ['9E', 'OO', 'ZW', 'XJ', 'WN', 'YV'] # Note the square brackets

#### Working with Lists
And just like strings (or any other sequence type) you can use slices:

In [23]:
print(airline_codes[::2])  # Starting with the first value, get ever other one

['9E', 'ZW', 'WN']


You can easily add to lists and remove from lists. Lists can also contain duplicate values.

In [24]:
airline_codes.append('WN')  # Appending a value to a list. It goes to the end.
print(airline_codes)

popped_value = airline_codes.pop()  # Let's remove the last value from the list and store it
print(popped_value)
print(airline_codes)

airline_codes.append([1,2,3])  # List within a lists...
print(airline_codes)

['9E', 'OO', 'ZW', 'XJ', 'WN', 'YV', 'WN']
WN
['9E', 'OO', 'ZW', 'XJ', 'WN', 'YV']
['9E', 'OO', 'ZW', 'XJ', 'WN', 'YV', [1, 2, 3]]


#### Iterating through Lists
I suppose I could have covered this with strings, but it wouldn't have been nearly as interesting.

Let's say we wanted to print out the type of objects within a list:

In [25]:
for i in airline_codes:
    print(type(i))

del airline_codes[-1]  # Remove the last value (the list) without returning the value

<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'list'>


I'll get into iterators more later, but just know that sequence type objects support iteration.

#### Checking for values
One last thing I'll demonstrate is checking for the existence of a value in a list using the `in` operator:

In [26]:
if 'DL' not in airline_codes:
    airline_codes.append('DL')
    print('Added DL to the list')

Added DL to the list


### Dictionaries
Dictionaries can be seen as storing key/value pairs. While lists are always indexed by integer, dictionaries can use any immutable type as a key (including tuples, as long as they themselves contain only immutable types - tuples will be covered later).

Dictionaries can be created empty, or created with initial values.

In [27]:
code_pairings = {}  # Create an empty dict
code_pairings['9E'] = 'Endeavor' # Assign a value for a given key

print(code_pairings)

code_pairings = {
    '9E': 'Endeavor', # Notice that when doing it this way, you use a colon
    'OO': 'SkyWest', # Also note you can have trailing commas
}

print(code_pairings)

{'9E': 'Endeavor'}
{'9E': 'Endeavor', 'OO': 'SkyWest'}


Dictionaries, like lists, can be iterated over. We can iterate over the keys, the keys and values, or just the values:

In [28]:
for k in code_pairings.keys():  # Just the keys
    print(k)

print('\n')

for k, v in code_pairings.items(): # Keys and values
    print('{key} contains {value}'.format(key=k, value=v))

print('\n')
    
for v in code_pairings.values(): # Just the values - although I've never actually used this before
    print(v)

9E
OO


9E contains Endeavor
OO contains SkyWest


Endeavor
SkyWest


There are other things you can do with dictionaries, but again, this is Lesson 1. I'm sure I'll show more advanced examples later.

### Tuples
The last thing we'll be covering in this lesson are tuples.

Just like lists, tuples are sequenced. However, unlike lists, tuples are immutable.

In [29]:
code_tuple = ('9E', 'OO', 'XJ')
print(code_tuple)
print(code_tuple[0])

try:
    del code_tuple[0]
except TypeError as err:
    print(err)

('9E', 'OO', 'XJ')
9E
'tuple' object doesn't support item deletion


You can also unpack tuples (well, lists as well, but I'm covering it now).

In fact, you've already seen an example of this when iterating through the key/value pairs of dictionaries.

In [30]:
x, y = (1, 2)
print(x)
print(y)

try:
    x, y = ('A', 'B', 'C')  # The value count has to match, though...
except ValueError as err:
    print(err)

1
2
too many values to unpack (expected 2)


And of course, you can iterate through tuples:

In [31]:
for x in (1, 2, 3, 4, 5):
    print(x)

1
2
3
4
5


One final thing about tuples: As I mentioned, they're immutable. However, the can contain objects which *are* mutable.

In [32]:
ex_tuple = (airline_codes, code_pairings)
print(ex_tuple)

del airline_codes[-1]
code_pairings['DL'] = 'Delta Air Lines'

print(ex_tuple)

(['9E', 'OO', 'ZW', 'XJ', 'WN', 'YV', 'DL'], {'9E': 'Endeavor', 'OO': 'SkyWest'})
(['9E', 'OO', 'ZW', 'XJ', 'WN', 'YV'], {'9E': 'Endeavor', 'OO': 'SkyWest', 'DL': 'Delta Air Lines'})


## Lesson Conclusion
I'm hoping by now you should have a good foundational understanding of working with Python. There are some things demo'd that I didn't cover yet, but that's ok. Play around. Have fun.

Also, don't be afraid to use this as a launching point start your own research and learning. I got my start with the language basically by coming up with my own task I wanted to accomplish and jumped from there (admittedly, I did have some prior experience with PHP).

I can also try to answer questions (and take feedback) left as an issue in the github repo.

Until next time.

-- Jason